### DISCUSSION TITLE
`checkForUpdate is not a function` in CircleCI

### DISCUSSION THREAD

**Discussion by knownasilya** [percy/cli#717] [Support] [answered]
## The problem

Running `percy exec -- cypress run` in CircleCI returns an error and crashes.

## Environment

- Node version: `14.2.0`
- `@percy/cli` version: `^1.0.0-beta.72`
- Version of Percy SDK you’re using: `"@percy/cypress": "^3.1.1"`
- If needed, a build or snapshot ID:
- OS version: CircleCI image `circleci/node:14.2.0-browsers`
- Type of shell command-line [interface]:

## Details

The error is 

```
(node:246) UnhandledPromiseRejectionWarning: TypeError: checkForUpdate is not a function
    at /home/circleci/teamgantt-web-client/node_modules/@percy/cli/bin/run:11:11
(Use `node --trace-warnings ...` to show where the warning was created)
(node:246) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:246) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

## Debug logs

Adding `--debug` didn't add any details, as this function seem to be early in the cli process.

**Comment by knownasilya** [comment, marked as answer]
`1.0.0-beta.71` seems to run but I get this

```
[percy] Stopping percy...
[percy] Build not created

Exited with code exit status 3
CircleCI received exit code 3
```

And 1 snapshot passed and is in the percy ui, but the 3 failed ones are missing.

```
Running:  components/my-projects/my-projects.js                                          (9 of 11)
[percy] Snapshot taken: My Projects can visit /my-projects

  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        4                                                                                │
  │ Passing:      1                                                                                │
  │ Failing:      3                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  3                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     17 seconds                                                                       │
  │ Spec Ran:     components/my-projects/my-projects.js                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /home/circleci/teamgantt-web-client/cypress/screenshots/components/my-projects/m     (1280x720)
     y-projects.js/My Projects -- can visit my-projectson-hold (failed).png                         
  -  /home/circleci/teamgantt-web-client/cypress/screenshots/components/my-projects/m     (1280x720)
     y-projects.js/My Projects -- can visit my-projectscomplete (failed).png                        
  -  /home/circleci/teamgantt-web-client/cypress/screenshots/components/my-projects/m     (1280x720)
     y-projects.js/My Projects -- can visit my-projectsother (failed).png
```

No additional details with `--debug` or `--verbose`


---

I had false positives locally because my local api server was running. That resolved the exit code 3 (which was just cypress erroring due to failed tests)

For the undefined error the issue was due to using version 2 of the circleci config. The v2 docs on percy need updating since they don't mention this. Found it here: https://docs.percy.io/v1/docs/circleci-orb-integration

**Reply by Robdel12** [reply]
Hey @knownasilya! Do you cache dependencies in CI? It sounds like there's caching / environmental issues going on -- pretty odd for both the flags to not produce verbose logs and for there to be a missing export. Or do you happen to have anything CLI related globally installed?

**Reply by knownasilya** [reply]
So I commented out the failing tests and got a 0 exit code. Weird thing is that when those failed snapshots (which pass locally) were there they don't end up on the percy.io site.

**Reply by Robdel12** [reply]
That's how Percy works -- if the test fails it likely doesn't make it to the `cy.percySnapshot` call. 

edit: are you still seeing an issue with `checkForUpdate is not a function`?

**Reply by knownasilya** [reply]
That issue is with the latest beta. I downgraded to the one before and don't see it. Will try revering back to latest in a bit.

On the failure this is a test:

```js
  it('can visit /my-projects/on-hold', () => {
    cy.visit('/my-projects/on-hold');
    cy.get('.my-projects .empty-list').should('exist');

    cy.percySnapshot();
  });
```

This fails on the snapshot and doesn't upload that failed image.
But if i remove the `cy.percySnapshot()` the tests pass just fine.

**Reply by Robdel12** [reply]
Yeah, it sounds like there's a caching issue going on. My bet is theres multiple different versions of the CLI  (its deps are lockstepped, so if any mismatch due to a bad cache things will break)

> doesn't upload that failed image.

FWIW, this isn't something Percy does. That's a Cypress thing

**Reply by knownasilya** [reply]
Ok, I figured out my issue with exit code 3, and it had to do with failing tests. Locally I forgot to turn off the api server which resulted in false positives.

For the `checkForUpdate is not a function` I resolved that with this doc I believe https://docs.percy.io/v1/docs/circleci-orb-integration. Note that this is a v1 doc and mentions CircleCI needing config 2.1 but the latest docs don't mention this.

**Reply by Robdel12** [reply]
Ah there we go, that makes sense. The orb isn't something I'd use (we pulled it from the orb registry a while ago), and likely the issue you're running into. That also has `@percy/agent` installed in the docker image, which will cause issues (two binaries named `percy`). I'd move away from the orb. If you're running parallel tests the new doc should get you going: https://docs.percy.io/docs/parallel-test-suites#parallel-suites-split-across-different-machines If you're not running parallel tests, the orb wasn't doing anything for ya :p

**Reply by istateside** [reply]
I had the same problem with `checkForUpdate is not a function` in our Drone configuration. I had to pin the version in our package.json to the previous beta version (71).

This ended up causing an issue, because the unhandled promise rejection didn't exit the process with `1`, so none of our test pipelines in drone were reporting as failed.

Would it be possible to add something like `process.on('unhandledRejection', (err) => process.exit(1))`, or catch potential rejections in bin/run.js to avoid this in the future?

**Reply by Robdel12** [reply]
Hey @istateside! The problem is `checkForUpdate is not a function` shouldn't even be an issue that occurs. That is only called in the latest version, so it should never be a missing export. Does your CI install a global version of `@percy/cli`? Or is there an old cache hanging around? What's likely happening is the global version is newer but there's also an older local version that gets imported. 

For example:

- beta 72: https://github.com/percy/cli/blob/v1.0.0-beta.72/packages/cli/bin/run
- beta 71: https://github.com/percy/cli/blob/v1.0.0-beta.71/packages/cli/bin/run

There has to be multiple versions hanging around 🤨

**Reply by knownasilya** [reply]
In my case there were not multiple versions. Unless percy/cypress added it's own cli version. This was the first time percy was added.

**Reply by istateside** [reply]
There must have been some NPM weirdness, but I can't say for sure exactly what was going wrong. I ran my tests with `npx @percy/cli exec -- npm run my-test-command`

I had to change that to `./node_modules/.bin/percy exec -- npm run my-test-command` to be sure it was using the right version

Looking at the code briefly, I agree that it _shouldn't_ ever happen, but it was definitely frustrating for us to go several hours of letting builds pass without realizing that none of them were running our cypress suite

**Reply by Robdel12** [reply]
I agree -- we originally wanted to support the ability to use a global install with a local package, but it appears `npx` caching is _really_ going to mess that up. I cut a new release that will always require its local CLI: https://github.com/percy/cli/releases/tag/v1.0.0-beta.73

This _shouldn't_ be an issue in beta.73.
